dhall/src/Dhall/Schemas.hs

```diff
@@ -0,0 +1,189 @@
+
+        Check -> do
+            if originalText == schemasText
+                then return ()
+                else do
+                    let command = "schemas"
+
+                    let modified = "rewritten"
+
+                    Exception.throwIO CheckFailed{..}
+
+decodeSchema
+    :: Expr Src Void
+    -> Maybe (Map Text (Expr Src Void), Map Text (Expr Src Void))
+decodeSchema
+    (RecordLit [ ("Type", Record _Type), ("default", RecordLit _default) ]) =
+    Just (_Type, _default)
+decodeSchema _ =
+    Nothing
+
+decodeSchemas
+    :: Expr Src Void
+    -> Maybe (Map Text (Map Text (Expr Src Void), Map Text (Expr Src Void)))
+decodeSchemas (RecordLit keyValues) = traverse decodeSchema keyValues
+decodeSchemas  _                    = Nothing
+
+-- | Simplify a Dhall expression using a record of schemas
+simplifyUsingSchemas
+    :: Expr Src Import
+    -- ^ Record of schemas
+    -> Expr Src Import
+    -- ^ Expression to simplify using the supplied schemas
+    -> IO (Expr Src Import)
+simplifyUsingSchemas _schemas expression = do
+    resolvedSchemas    <- Import.load _schemas
+    resolvedExpression <- Import.load expression
+
+    _ <- Core.throws (TypeCheck.typeOf resolvedSchemas)
+    _ <- Core.throws (TypeCheck.typeOf resolvedExpression)
+
+    let normalizedSchemas    = Normalize.normalize resolvedSchemas
+    let normalizedExpression = Normalize.normalize resolvedExpression
+
+    decodedSchemas <- case decodeSchemas normalizedSchemas of
+        Just decodedSchemas -> return decodedSchemas
+        Nothing             -> Exception.throwIO NotASchemaRecord
+
+    let schemasBinding value = Binding{..}
+          where
+            variable   = "schemas"
+            annotation = Nothing
+
+            bindingSrc0 = Nothing
+            bindingSrc1 = Nothing
+            bindingSrc2 = Nothing
+
+    let schemasRewrite subExpression =
+            (Maybe.fromMaybe subExpression . Maybe.listToMaybe) $ do
+                (name, (_Type, _default)) <- Map.toList decodedSchemas
+
+                defaultedRecord <- case subExpression of
+                    RecordLit keyValues -> do
+                        let defaultedKeyValues =
+                                Map.fromList (Map.toList keyValues \\ Map.toList _default)
+
+                        return (RecordLit defaultedKeyValues)
+
+                    _ -> empty
+
+                case TypeCheck.typeOf (Annot (Let (schemasBinding resolvedSchemas) subExpression) (Record _Type)) of
+                    Left  _ -> empty
+                    Right _ -> return ()
+
+                return (RecordCompletion (Field "schemas" name) defaultedRecord)
+
+    let rewrittenExpression =
+            fmap Void.absurd (Optics.transformOf Syntax.subExpressions schemasRewrite normalizedExpression)
+
+    let usesSchema = (not . null) $ do
+            Var (V "schemas" 0) <- Optics.universeOf Syntax.subExpressions rewrittenExpression
+            return ()
```

Comment:
(Aside: I asked on Slack how to solve this with `anyOf` and got some help from @danidiaz:

```suggestion
    let usesSchema =
            Lens.Family.anyOf
                (Optics.cosmosOf Syntax.subExpressions)
                (\case
                    Var (V "schemas" 0) -> True
                    _ -> False
                )
                rewrittenExpression
```

where `cosmosOf` is

```haskell
cosmosOf :: (Applicative f, Contravariant f) => LensLike' f a a -> LensLike' f a a
cosmosOf d f s = f s *> d (cosmosOf d f) s
```

This isn't necessarily better, but it does avoid the somewhat accidental (but admittedly elegant) use of the list.

Since we already use `anyOf` in a somewhat convoluted way in other parts of `dhall`, maybe we can standardize on a version of one of these patterns

https://github.com/dhall-lang/dhall-haskell/blob/0008cd03d21ea7ce52a74983d663c3cdd34a32ff/dhall/src/Dhall/TypeCheck.hs#L962-L984

)

Reply:
I switched to `cosmosOf` + `anyOf` as you suggested

The one minor thing I wanted to point out is that `cosmosOf` does not require a `Contravariant` constraint.  It's not clear to me why `lens` adds that constraint

Reply:
Thanks! I've brought it up in https://github.com/ekmett/lens/pull/929.